### PR TITLE
libroach: add missing destructor (in parent class)

### PR DESCRIPTION
**Pull request by knz**
I think this is slightly better future-proof (you can carry `unique_ptr<KeyManager>` with this)

### CODE REVIEW

**Comment by petermattis** [comment]
:lgtm:

This is good C++ style. As a reminder for others, any C++ class which declares virtual functions should declare a virtual destructor (or inherit from a class which declares a virtual destructor). I don't recall all of the badness that results from not doing so, but recall that it is there. Something about the destructor not being called properly. 

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21497#-:-L30alv-Yu67A2BvKesZ:b-dn23zg)*

File: c-deps/libroach/ccl/key_manager.h [hunk 1]
```diff
@@ -85,8 +87,6 @@ class FileKeyManager : public KeyManager {
                           const std::string& old_key_path)
       : env_(env), active_key_path_(active_key_path), old_key_path_(old_key_path) {}
 
-  virtual ~FileKeyManager();
```
Comment (mberhault) [inline c-deps/libroach/ccl/key_manager.h]: That should stay for any direct users of `FileKeyManager`

**Comment by mberhault** [comment]
If the abstract class does not have a virtual destructor, then anyone holding a pointer to the abstract class will not call the derived class' destructor (I'm not doing that yet, but I will soon). We still want the virtual destructor on the derived class though.

File: c-deps/libroach/ccl/key_manager.h (hunk 1 above)
Comment (knz) [reply c-deps/libroach/ccl/key_manager.h]: No it doesn't: the parent class of `FileKeyManager` already has a virtual destructor (with my change above), so automatically all derived classes have their destructor virtual too.

Comment (mberhault) [reply c-deps/libroach/ccl/key_manager.h]: Sure, but it's still weird to have `~FileKeyManager()` in the `.cc` that isn't reflected in the `.h`.

Comment (knz) [reply c-deps/libroach/ccl/key_manager.h]: oh I see, sorry for that. I am so used to template programming with all code in headers I hadn't even looked at the cc file. 

Fixed, PTAL.

**Comment by mberhault** [comment]
LGTM. Thanks for the fix. Small nit: you can leave the actual implementation (`~KeyManager(){}`) in the `.h`, the reason to avoid inline destructors is when they may be too big, but `KeyManager` doesn't have anything to cleanup.

**Comment by knz** [comment]
> Small nit: you can leave the actual implementation (~KeyManager(){}) in the .h, the reason to avoid inline destructors is when they may be too big

The same argument would apply to `~FileKeyManager()`, doesn't it? If I do it for one then the other should be the same for consistency.

In general if the class is not a template i'd rather have the definition in a cc file to avoid linker churn. The churn is the same regardless of function body size.

**Comment by knz** [comment]
The argument about inlining just makes me think...

I now realize there is a bunch of `virtual` methods on *non-template* classes in the header file. AFAIK virtual methods cannot be inlined in most cases. I do not see that there is a benefit to have their definition in the header file (again, non-template). In fact I think it adds linker churn without any upside. What do you think?

**Comment by knz** [comment]
I appended 2 commits to the PR to clarify what I mean. PTAL.

File: c-deps/libroach/ccl/key_manager.h [hunk 2]
```diff
@@ -74,6 +74,8 @@ class KeyManager {
   // GetKey returns the key with requested `id`.
   // **WARNING**: this must not be logged, displayed, or stored outside the key registry.
   virtual std::unique_ptr<enginepbccl::SecretKey> GetKey(const std::string& id) = 0;
+
+  virtual ~KeyManager();
```
Comment (mberhault) [inline c-deps/libroach/ccl/key_manager.h]: tiny style nit: the destructor declaration should be at the beginning of the `public:` section.

**Comment by mberhault** [comment]
LGTM. Fair enough.

File: c-deps/libroach/ccl/key_manager.h (hunk 2 above)
Comment (knz) [reply c-deps/libroach/ccl/key_manager.h]: Done

**Pull request merged**